### PR TITLE
script with aliases and new id submission

### PR DESCRIPTION
**Pull request by Antichess**

### CODE REVIEW

File: mainscript.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        counts = walk_thread(get_comment)[::-1]
+        base_count = find_count_in_text(counts[0][0])
+        thread_duration = datetime.timedelta(seconds=counts[-1][2] - counts[0][2])
+        days = thread_duration.days
+        hours, mins, secs = str(thread_duration).split(':')[-3:]
+     
+        counters = defaultdict(int)
+
+        aliaseslist = []
+        f = open((os.getcwd()+"\input\prefs\Aliases.txt"),"r")
```
Comment (cutonbuminband) [inline mainscript.py]: See point 3

File: mainscript.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        try:
+            context = (input().split("/"))[8]
+            print(context)
+            ids.append(context)
+        except:
+            break
+    
+    for current_id in ids:
+        t_start = datetime.datetime.now()
+        reddit_instance = praw.Reddit(client_id=client_id,
```
Comment (cutonbuminband) [inline mainscript.py:81]: See point 2

File: mainscript.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import praw
+import os
+
+#the ids are now in a list (array). much easier to do 
+#and you can have it idle in the background
+
+client_id="14 char client secret"
+client_secret="30 char client secret"
+ 
+def find_count_in_text(body):
```
Comment (cutonbuminband) [inline mainscript.py:15]: See point 1

File: mainscript.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        regex = ("^[^\d]*"    # We strip non-numeric characters from the start
+                 "([\d, \.]*)")  # And then we take all digits and separators
+        count = re.findall(regex, body)[0]
+        # We remove any separators, and try to convert the remainder to an int.
+        stripped_count = count.translate(str.maketrans('', '', ' ,.'))
+        return int(stripped_count)
+    except:
+        raise ValueError(f"Unable to extract count from comment body: {body}")
+ 
+def comment_to_tuple(comment):
```
Comment (cutonbuminband) [inline mainscript.py:26]: See point 1

File: mainscript.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    author = str(comment.author) if comment.author is not None else "[deleted]"
+    return (comment.body, author, comment.created_utc, comment.id, str(comment.submission))
+ 
+def walk_thread(leaf_comment):
```
Comment (cutonbuminband) [inline mainscript.py:30]: See point 1

File: mainscript.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+     
+        counters = defaultdict(int)
+
+        aliaseslist = []
+        f = open((os.getcwd()+"\input\prefs\Aliases.txt"),"r")
+        lines = f.readlines()
+        for line in lines:
+            aliaseslist.append(line.replace("\n","",1).split(","))
+        
+        with open(os.getcwd()+"\\results\\LOG_"+str(base_count)+"to"+str(base_count+1000)+".csv", "w") as hog_file:
```
Comment (cutonbuminband) [inline mainscript.py]: See point 3

File: mainscript.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                if replacealt is False:
+                    counters['/u/' + x[1]] += 1
+
+                #print(str(x[0]) + str(x[1]) + str(x[2]) + str(x[3]))
+                try:
+                    print(base_count + idx + 1, *x[1:], sep=",", file=hog_file)
+                except:
+                    continue
+     
+        with open(os.getcwd()+"\\results\\TABLE_"+str(base_count)+"to"+str(base_count+1000)+".csv", "w") as hoc_file:
```
Comment (cutonbuminband) [inline mainscript.py]: See point 3

File: mainscript.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                print(rank + 1, *counter, sep="|", file=hoc_file)
+     
+            hoc_file.write(f"\nIt took {rank+1} counters "
+                           f"{days} days {hours} hours {mins} mins {secs} secs "
+                           "to complete this thread. Bold is the user with the get"
+                           f"\ntotal counts in this chain logged: {total_counts}")
+     
+        elapsed_time = datetime.datetime.now() - t_start
+        print(elapsed_time)
+        print(os.getcwd()+"\\results\\LOG_"+str(base_count)+"to"+str(base_count+1000)+".csv")
```
Comment (cutonbuminband) [inline mainscript.py]: See point 3

File: mainscript.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        print(os.getcwd()+"\\results\\TABLE_"+str(base_count)+"to"+str(base_count+1000)+".csv")
```
Comment (cutonbuminband) [inline mainscript.py]: See point 3

File: mainscript.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+        aliaseslist = []
+        f = open((os.getcwd()+"\input\prefs\Aliases.txt"),"r")
+        lines = f.readlines()
+        for line in lines:
+            aliaseslist.append(line.replace("\n","",1).split(","))
+        
+        with open(os.getcwd()+"\\results\\LOG_"+str(base_count)+"to"+str(base_count+1000)+".csv", "w") as hog_file:
+            for idx, x in enumerate(counts[1:]):
+                replacealt = False
```
Comment (cutonbuminband) [inline mainscript.py]: See point 4

File: mainscript.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+     
+        with open(os.getcwd()+"\\results\\TABLE_"+str(base_count)+"to"+str(base_count+1000)+".csv", "w") as hoc_file:
+            hoc_file.write("Thread Participation Chart for " + title + "\n\n")
+            hoc_file.write("Rank|Username|Counts\n")
+            hoc_file.write("---|---|---\n")
+            counters = sorted(counters.items(), key=lambda kv: (kv[1], kv[0]))[::-1]
+            total_counts = sum([x[1] for x in counters])
+     
+            for rank, counter in enumerate(counters):
+                if counter[0][3:] == str(get_comment.author):
```
Comment (cutonbuminband) [inline mainscript.py]: See point 4

**Review by cutonbuminband — CHANGES_REQUESTED** [review]
Hey Anti,

Awesome that you're helping make these scripts better! I will definitely be merging this pull request, but before I do, I have some comments, and a few things that need to be taken care of. That's perfectly normal when coding collaboratively, both at work and in open source hobby coding.

I could make the changes myself, and if you would like me to I will. But before I do that, I think you should have a look yourself.

1. I think it's a good thing that you made your mainscript.py a new file - it avoids conflicts with the existing code. It looks like you've based the file on the one I put in the pastebin, which I pulled out as a standalone script. That means that some of the functions you've defined in the code already exist in other files. The closest file to `mainscript.py` in the repository is `log_thread.py`. Take a look at that one and see how it imports two functions from other files in the repository. Specifically the lines `from parsing import find_count_in_text` and `from thread_navigation import walk_thread` mean that I didn't have to recode those functions in this file. Could you go through `mainscript.py` and do the same thing?

2. For interacting with reddit, I have two comments:
   - If you look at `log_thread.py`, you'll see that the reddit instance is initialised with `reddit_instance = praw.Reddit("stats_bot")`, instead of explicitly passing in credentials. That's because I have put all the praw configuration and secrets in a file called `praw.ini`, as described [here](https://github.com/cutonbuminband/counting_stats#installation-and-usage), and made sure that file is never committed to git by adding a line in the file `.gitignore`. I strongly suggest you do the same: it lets us easily separate the secret and not-secret parts of the code.
   - At the moment, a new reddit instance is being created for every thread, because that line is inside the loop. But there's no need for that: we can just reuse the same one over again.

3. This program has now gone cross-platform! That's pretty cool, given that it only has two users. Unfortunately, this means that we need to do some work, since Windows uses "\" as a path separator, and linux uses "/". Luckily, there exists a library to take care of that for us called pathlib. If you add a line to the top of the file that says `from pathlib import Path`, and change things like `(os.getcwd()+"\input\prefs\Aliases.txt"` to `Path(os.getcwd()) / Path("input/prefs/Aliases.txt"`, everything should work on my machine as well. All the slashes in the new version should be forward slashes, and then pathlib automatically picks the right version for windows and linux.

4. I'm really glad you've included the aliases list. It was something I'd completely forgotten about! I think the best way to use it would be to make a function (called something like `canonical_username`) that takes as input a username and a the aliases, and returns either the original username if there are no aliases, or the canonical name if there are. That way, we can easily get the true form of any username. In particular, we can use it to fix a bug on line 112, where the result is currently wrong if the get_author is an alias.

**What happens now?**

If you are OK with trying to make the changes above, or just some of them, you should do that, and once you are happy with them, push them to your github repository. This pull request will then be automatically updated. As I wrote, you can do as much as as little as you want - and if you get stuck, give me a shout. Once you are finished, I'll make any fixes that are still necessary, and merge the pull request. If you feel I'm being too slow about it, ping me, and I'll get back to you within 48h.

**Comment by Antichess** [comment]
Alright. I'm confused and tried to fix a few things, so it would be nice if we could talk over Discord. I've sent a friend request, and it'd be nice if we could talk there.

**Pull request merged**